src/buffer.ts

```diff
@@ -0,0 +1,21 @@
+const { assign } = Object;
+
+export default class Buffer {
+  private target: object;
+  private changes: object;
+  [propName: string]: any;
+
+  constructor(target: object) {
+    this.target = target;
+    this.changes = {};
+  }
+
+  public set<T>(key: PropertyKey, value: T): T {
```

Comment:
The Buffer may need to have some kind of internal proxy as well so we can trap sets